src/git/git.go

```diff
@@ -0,0 +1,149 @@
+package git
+
+import (
+	"errors"
+	"os"
+	"strings"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing/transport/ssh"
+)
+
+const (
+	colorCyan   = "\033[36m"
+	colorYellow = "\033[33m"
+	colorReset  = "\033[0m"
+)
+
+type Logger interface {
+	Info(s string, args ...interface{})
+	Error(s string, args ...interface{})
+	Warn(s string, args ...interface{})
+}
+
+type ElapsedTime func(functionName string) func()
+
+type GitVersioning struct {
+	log                  Logger
+	printElapsedTime     ElapsedTime
+	url                  string
+	destinationDirectory string
+	sshKey               string
+	replace              bool
+	repo                 *git.Repository
+	authorization        *ssh.PublicKeys
+}
+
+// GetPublicKey aims to read ssh key file and returns a go git PublicKey
+// Args:
+// 		sshKey ([]byte): Byte arry containing the ssh key.
+// 		Note: Set up this key in the git settings and pass as a parameter to the CLI tool.
+// Returns:
+// 		error: Returns an error whenever unexpected issues happen.
+func (g *GitVersioning) getPublicKey(sshKey []byte) error {
+	var publicKey *ssh.PublicKeys
+
+	publicKey, err := ssh.NewPublicKeys("git", []byte(sshKey), "")
+	if err != nil {
+		g.log.Error(err.Error())
+		return err
+	}
+
+	g.authorization = publicKey
+	return nil
+}
+
+// FormatSSHKey aims to properlly format the ssh private key.
+// It is necessary once we cannot pass a multiline variable to go build. So it is not possible
+// to do docker build --build-arg PRIVATE_KEY=$(PRIVATE_KEY) once this kind of file has several line breaks.
+// Args:
+// 		lineBreaker (string): Line break holder character. I.e.: #
+// Returns:
+// 		string: Returns the sshKey string with lineBreaker character replaced by line breaks.
+func (g *GitVersioning) formatSSHKey(lineBreaker string) string {
+	return strings.Replace(g.sshKey, lineBreaker, "\n", -1)
+}
+
+func (g *GitVersioning) validate() error {
+	if g.url == "" {
+		return errors.New("url cannot be empty")
+	}
+
+	if g.destinationDirectory == "" {
+		return errors.New("destination directory cannot be empty")
+	}
+
+	if g.sshKey == "" {
+		return errors.New("ssh key cannot be empty")
+	}
+
+	return nil
+}
+
+// cloneRepoToDirectory aims to clone the repository from remote to local.
+// Args:
+// 		auth (*ssh.PublicKeys): Authorization key.
+// Returns:
+// 		*git.Repository: Returns a repository reference.
+// 		err: Error whenever unexpected issues happen.
+func (g *GitVersioning) cloneRepoToDirectory() (*git.Repository, error) {
+	defer g.printElapsedTime("CloneRepoToDirectory")()
+
+	if err := g.validate(); err != nil {
+		g.log.Error(err.Error())
+		return nil, err
+	}
+
+	err := g.getPublicKey([]byte(g.formatSSHKey("#")))
+	if err != nil {
+		g.log.Error("error while getting public key due to: %s", err)
+		return nil, err
+	}
+
+	g.log.Info(colorYellow+"cloning repo "+colorCyan+" %s "+colorYellow+" into "+colorCyan+"%s"+colorReset, g.url, g.destinationDirectory)
```

Comment:
That would be a little hard to understand where color starts and ends.

Reply:
That's the only reason why I did not use fmt.Sprintf here.

Unlike python where we can use f-string like `f"this is {blue_color}blue{reset_color}"`, in go it can be hard to read, because it works like format function.

Imagine an hipotetic case in which we have this:

`fmt.Sprintf("%ssomething went wrong due to %s%v%s", yellowColor, redColor, err.Error(), resetColor)`

I think it is hard to read and understand where each color starts and ends.

What do you think?